File: source/swissimage/cache.py
import sqlite3
from contextlib import closing
from typing import Optional, Tuple, List
DB_PATH = "./cache/swissimage.sqlite3"

initialized = False

def _get_connection():
    return sqlite3.connect(DB_PATH)


def initialize_cache():
    global initialized
    if not initialized:
        with closing(_get_connection()) as conn:
            c = conn.cursor()
            c.execute('''
                CREATE TABLE IF NOT EXISTS swissimage_references (
                    id TEXT NOT NULL,
                    modify_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                    PRIMARY KEY (id)
                )
            ''')
            c.execute('''
                CREATE TABLE IF NOT EXISTS swissimage_data (
                    x INTEGER NOT NULL,
                    y INTEGER NOT NULL,
                    r INTEGER NOT NULL,
                    g INTEGER NOT NULL,
                    b INTEGER NOT NULL,
                    reference_id TEXT,
                    modify_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                    PRIMARY KEY (x, y),
                    FOREIGN KEY (reference_id) REFERENCES swissimage_references(id)
                )
            ''')
            conn.commit()
            initialized = True

def get_from_cache(x: int,y: int)->Optional[Tuple[int, int, int]]:
    with closing(_get_connection()) as conn:
        c = conn.cursor()
        c.execute('SELECT r, g, b FROM swissimage_data WHERE x = ? AND y = ?', (x,y))
        data = c.fetchone()
        return (int(data[0]),int(data[1]),int(data[2])) if data is not None else None

def get_many_from_cache(reference: str)-> Optional[List[Tuple[int,int,int,int,int]]]:
    with closing(_get_connection()) as conn:
        c = conn.cursor()
        c.execute('SELECT x, y, r, g, b FROM swissimage_data WHERE reference_id IS ?', (reference,))
        data = c.fetchall()
        return data if len(data) > 0 else None

def get_many_from_cache_filtered(**kwargs)-> Optional[List[Tuple[int,int,int,int,int]]]:
    step = kwargs['step'] or 0;
    minx = kwargs['minx'];
    maxx = kwargs['maxx'];
    miny = kwargs['miny'];
    maxy = kwargs['maxy'];

    with closing(_get_connection()) as conn:
        c = conn.cursor()
        if None not in [minx, maxx, miny, maxy]:
            c.execute('''
                SELECT x,y,r,g,b FROM swissimage_data
                WHERE
                        x < ?
                    AND x > ?
                    AND y < ?
                    AND y > ?

                ORDER BY x,y ASC
                ''', (maxx, minx, maxy, miny))
        else:
            c.execute('''
                SELECT x,y,r,g,b FROM swissimage_data
                WHERE
                    (x % ? = 0) AND (y % ? = 0)
                ORDER BY x,y ASC
                ''', (step, step))
        data = c.fetchall()
    return data if len(data) > 0 else None



def write_to_cache(x: int, y: int, color: Tuple[int,int,int], **kwargs):
    reference = str(kwargs.get('reference', None));
    with closing(_get_connection()) as conn:
        c = conn.cursor()
        if reference is not None:
            c.execute('INSERT OR REPLACE INTO swissimage_references (id) VALUES (?)', (reference,))
        c.execute('INSERT OR REPLACE INTO swissimage_data (x, y, r, g, b, reference_id) VALUES (?, ?, ?, ?, ?, ?)', (x,y,*color, reference))
        conn.commit()

def write_many_to_cache(data: List[Tuple[Tuple[int,int],Tuple[int,int,int]]], reference: str):
    with closing(_get_connection()) as conn:
        c = conn.cursor()
        c.execute('INSERT OR REPLACE INTO swissimage_references (id) VALUES (?)', (reference,))
        c.executemany('INSERT OR REPLACE INTO swissimage_data (x, y, r, g, b, reference_id) VALUES (?, ?, ?, ?, ?, ?)',
            ((x, y, int(r), int(g), int(b), reference) for ((x, y), (r,g,b)) in data))
        conn.commit()

File: rust/src/swissimage/cache.rs
use std::path::Path;
use std::{fs, time::Duration};
use tokio_rusqlite::{params, Connection, Result};

use crate::utils::bounding_box::BoundingBox;

const SWISSIMAGE_CACHE_FILE: &str = "cache/swissimage.sqlite3";

pub async fn initialize_cache() -> Result<Connection> {
    let db_path = Path::new(SWISSIMAGE_CACHE_FILE);
    // create directories recursively if not exist
    if let Some(parent_dir) = db_path.parent() {
        let _ = fs::create_dir_all(parent_dir);
    }
    let conn = Connection::open(SWISSIMAGE_CACHE_FILE).await?;
    conn.call(|conn| {
        conn.busy_timeout(Duration::from_millis(30000))?;
        Ok(conn.execute_batch(
            "
            BEGIN;
            -- References
            CREATE TABLE IF NOT EXISTS swissimage_references (
                ref_id INTEGER PRIMARY KEY AUTOINCREMENT,
                name TEXT NOT NULL,
                modify_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            );

            -- R*Tree table for fast queries
            CREATE VIRTUAL TABLE IF NOT EXISTS swissimage_data
            USING rtree(
                rowid,
                x_min, x_max,
                y_min, y_max,
                +r INTEGER NOT NULL,
                +g INTEGER NOT NULL,
                +b INTEGER NOT NULL,
                +reference_id INTEGER
            );

            COMMIT;
            ",
        ))
    })
    .await??;

    Ok(conn)
}

pub async fn get_from_cache(
    datapoint_percent: f64,
    bounding_box: &BoundingBox,
) -> Result<Vec<(i32, i32, u8, u8, u8)>> {
    let bounding_box = (*bounding_box).clone();
    let conn = initialize_cache().await?;
    let rows = conn
        .call(move |conn| {
            let mut stmt = conn.prepare(
                "
                SELECT * FROM (
                    SELECT x_min as x, y_min as y, r, g, b, rowid
                        FROM swissimage_data
                        WHERE
                            x_min BETWEEN ?2 AND ?3 AND -- bounding box x
                            y_min BETWEEN ?4 AND ?5     -- bounding box y                
                    )
                WHERE rowid % ?1 == 0
                ORDER BY x,y;                
            ",
            )?;
            let mut rows = stmt.query((
                100f64 / datapoint_percent,
                bounding_box.x_range.0,
                bounding_box.x_range.1,
                bounding_box.y_range.0,
                bounding_box.y_range.1,
            ))?;
            let mut parsed_rows: Vec<(i32, i32, u8, u8, u8)> = vec![];
            while let Some(row) = rows.next()? {
                let x: f64 = row.get("x")?;
                let y: f64 = row.get("y")?;
                let r: u8 = row.get("r")?;
                let g: u8 = row.get("g")?;
                let b: u8 = row.get("b")?;
                parsed_rows.push((x as i32, y as i32, r, g, b));
            }
            Ok(parsed_rows)
        })
        .await?;
    Ok(rows)
}

pub async fn write_to_cache(data: &Vec<(u32, u32, u8, u8, u8)>, reference: &str) -> Result<()> {
    let data = data.clone();
    let reference = reference.to_string();
    let conn = initialize_cache().await?;
    conn.call(move |conn|{
        let tx = conn.transaction()?;
        {
            tx.execute(
                "INSERT OR REPLACE INTO swissimage_references (name) VALUES (?1)",
                [reference],
            )?;
            let ref_id = tx.last_insert_rowid();
            let mut stmt = tx.prepare(
                "INSERT OR REPLACE INTO swissimage_data (x_min, x_max, y_min, y_max, r, g, b, reference_id) VALUES (?1, ?1, ?2, ?2, ?3, ?4, ?5, ?6)",
            )?;
            for (x, y, r, g, b) in data {
                stmt.execute(params![x, y, r, g, b, ref_id])?;
            }
        }
        tx.commit()?;
        Ok(())
    }).await
}

pub async fn check_cache(reference: &str) -> Result<bool> {
    let conn = initialize_cache().await?;
    let reference = reference.to_string();
    conn.call(|conn| {
        let count_found: isize = conn.query_row(
            "SELECT COUNT(*) FROM swissimage_references WHERE name IS ?1",
            [reference],
            |row| row.get(0),
        )?;

        Ok(count_found > 0)
    })
    .await
}
